examples/petsc/ex28p.cpp

```diff
@@ -0,0 +1,388 @@
+   //    this example we do 'ref_levels' of uniform refinement. We choose
+   //    'ref_levels' to be the largest number that gives a final mesh with no
+   //    more than 10,000 elements.
+   for (int lev = 0; lev < ser_ref_levels; lev++)
+   {
+      mesh->UniformRefinement();
+   }
+
+   // 5. Define a parallel mesh by a partitioning of the serial mesh. Refine
+   //    this mesh further in parallel to increase the resolution. Once the
+   //    parallel mesh is defined, the serial mesh can be deleted.
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int l = 0; l < par_ref_levels; l++)
+   {
+      pmesh->UniformRefinement();
+   }
+   pmesh->ReorientTetMesh();
+
+   // 6. Define a parallel finite element space on the parallel mesh. Here we
+   //    use the Nedelec finite elements of the specified order.
+   std::cout << "dim: " << dim << "\n";
+   FiniteElementCollection *hcurl_coll = new ND_FECollection(order, dim);
+   FiniteElementCollection *h1_coll = new H1_FECollection(order, dim);
+
+   ParFiniteElementSpace *N_space = new ParFiniteElementSpace(pmesh, hcurl_coll);
+   ParFiniteElementSpace *L_space = new ParFiniteElementSpace(pmesh, h1_coll);
+
+   HYPRE_Int dimN = N_space->GlobalTrueVSize();
+   HYPRE_Int dimL = L_space->GlobalTrueVSize();
+
+   if (verbose)
+   {
+      std::cout << "***********************************************************\n";
+      std::cout << "dim(N) = " << dimN << "\n";
+      std::cout << "dim(L) = " << dimL << "\n";
+      std::cout << "dim(N+L) = " << dimN + dimL << "\n";
+      std::cout << "***********************************************************\n";
+   }
+
+   // 7. Define the two BlockStructure of the problem.  block_offsets is used
+   //    for Vector based on dof (like ParGridFunction or ParLinearForm),
+   //    block_trueOffstes is used for Vector based on trueDof (HypreParVector
+   //    for the rhs and solution of the linear system).  The offsets computed
+   //    here are local to the processor.
+   Array<int> block_offsets(3); // number of variables + 1
+   block_offsets[0] = 0;
+   block_offsets[1] = N_space->GetVSize();
+   block_offsets[2] = L_space->GetVSize();
+   block_offsets.PartialSum();
+
+   Array<int> block_trueOffsets(3); // number of variables + 1
+   block_trueOffsets[0] = 0;
+   block_trueOffsets[1] = N_space->TrueVSize();
+   block_trueOffsets[2] = L_space->TrueVSize();
+   block_trueOffsets.PartialSum();
+
+   // 8. Define the coefficients of the PDE.
+   ConstantCoefficient u_r_func(1.0);
+   Vector e_r(2);
+   double k0 = M_PI*2/1.0;
+   // We lump the sign, the wavenumber and the dielectric constant into one
+   // coefficient. The dielectric contsant is the square of the refractive
+   // index.
+   e_r(0) = -pow(k0*1.0,2);
+   // This is an example to use different refractive indices in mesh domains
+   e_r(1) = -pow(k0*2.0,2);
+   PWConstCoefficient e_r_func(e_r);
```

Comment:
I believe the default mesh has only one material region so this will always return `e_r(0)`.  It would be good to modify the mesh attributes to create a dielectric slab configuration.  Otherwise this is not really a dielectric waveguide but just a metallic waveguide.

Reply:
Yes, with only one material this is simulating a metallic waveguide. Should I add an example mesh with two material regions that would resemble a dielectric waveguide? I could only find e.g. `beam-tri.mesh` which is not a very useful example of dielectric waveguide.

Reply:
It might be easier to reassign selected element attributes like this:
```
   {
      Vector cent(dim);
      for (int i=0; i<pmesh->GetNE(); i++)
      {
         pmesh->GetElementCenter(i, cent);
	 if (fabs(cent[0]-0.5) < 0.125)
	 {
	   pmesh->GetElement(i)->SetAttribute(2);
	 }
      }
   }
```

Reply:
I added something similar in https://github.com/mfem/mfem/pull/1487/commits/d27f3d683e041b3af4c488f1ad3a818185a8b09a to define a rectangular waveguide core.